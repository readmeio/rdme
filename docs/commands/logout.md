`rdme logout`
=============

Logs the currently authenticated user out of ReadMe.

* [`rdme logout`](#rdme-logout)

## `rdme logout`

Logs the currently authenticated user out of ReadMe.

```
USAGE
  $ rdme logout

DESCRIPTION
  Logs the currently authenticated user out of ReadMe.
```
