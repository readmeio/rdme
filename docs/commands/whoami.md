`rdme whoami`
=============

Displays the current user and project authenticated with ReadMe.

* [`rdme whoami`](#rdme-whoami)

## `rdme whoami`

Displays the current user and project authenticated with ReadMe.

```
USAGE
  $ rdme whoami

DESCRIPTION
  Displays the current user and project authenticated with ReadMe.
```
