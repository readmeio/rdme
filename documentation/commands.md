# Table of contents

  <!-- toc -->
* [Table of contents](#table-of-contents)
* [Example Usage](#example-usage)
* [Command Reference](#command-reference)
<!-- tocstop -->

# Example Usage

  <!-- usage -->
```sh-session
$ npm install -g rdme
$ rdme COMMAND
running command...
$ rdme (--version)
rdme/9.0.0-next.21 darwin-arm64 node-v22.11.0
$ rdme --help [COMMAND]
USAGE
  $ rdme COMMAND
...
```
<!-- usagestop -->

# Command Reference

  <!-- commands -->
* [`rdme autocomplete [SHELL]`](#rdme-autocomplete-shell)
* [`rdme categories`](#rdme-categories)
* [`rdme categories:create TITLE`](#rdme-categoriescreate-title)
* [`rdme changelogs PATH`](#rdme-changelogs-path)
* [`rdme custompages PATH`](#rdme-custompages-path)
* [`rdme docs PATH`](#rdme-docs-path)
* [`rdme docs:prune FOLDER`](#rdme-docsprune-folder)
* [`rdme help [COMMAND]`](#rdme-help-command)
* [`rdme login`](#rdme-login)
* [`rdme logout`](#rdme-logout)
* [`rdme open`](#rdme-open)
* [`rdme openapi [SPEC]`](#rdme-openapi-spec)
* [`rdme openapi:convert [SPEC]`](#rdme-openapiconvert-spec)
* [`rdme openapi:inspect [SPEC]`](#rdme-openapiinspect-spec)
* [`rdme openapi:reduce [SPEC]`](#rdme-openapireduce-spec)
* [`rdme openapi:validate [SPEC]`](#rdme-openapivalidate-spec)
* [`rdme versions`](#rdme-versions)
* [`rdme versions:create VERSION`](#rdme-versionscreate-version)
* [`rdme versions:delete [VERSION]`](#rdme-versionsdelete-version)
* [`rdme versions:update [VERSION]`](#rdme-versionsupdate-version)
* [`rdme whoami`](#rdme-whoami)

## `rdme autocomplete [SHELL]`

Display autocomplete installation instructions.

```
USAGE
  $ rdme autocomplete [SHELL] [-r]

ARGUMENTS
  SHELL  (zsh|bash|powershell) Shell type

FLAGS
  -r, --refresh-cache  Refresh cache (ignores displaying instructions)

DESCRIPTION
  Display autocomplete installation instructions.

EXAMPLES
  $ rdme autocomplete

  $ rdme autocomplete bash

  $ rdme autocomplete zsh

  $ rdme autocomplete powershell

  $ rdme autocomplete --refresh-cache
```

_See code: [@oclif/plugin-autocomplete](https://github.com/oclif/plugin-autocomplete/blob/v3.2.8/src/commands/autocomplete/index.ts)_

## `rdme categories`

Get all categories in your ReadMe project.

```
USAGE
  $ rdme categories --key <value> [--version <value>]

FLAGS
  --key=<value>      (required) Project API key
  --version=<value>  Project version. If running command in a CI environment and this option is not passed, the main
                     project version will be used.

DESCRIPTION
  Get all categories in your ReadMe project.
```

## `rdme categories:create TITLE`

Create a category with the specified title and guide in your ReadMe project.

```
USAGE
  $ rdme categories:create TITLE --categoryType guide|reference --key <value> [--preventDuplicates] [--version <value>]

ARGUMENTS
  TITLE  Title of the category

FLAGS
  --categoryType=<option>  (required) Category type
                           <options: guide|reference>
  --key=<value>            (required) Project API key
  --preventDuplicates      Prevents the creation of a new category if there is an existing category with a matching
                           `categoryType` and `title`
  --version=<value>        Project version. If running command in a CI environment and this option is not passed, the
                           main project version will be used.

DESCRIPTION
  Create a category with the specified title and guide in your ReadMe project.
```

## `rdme changelogs PATH`

Sync Markdown files to your ReadMe project as Changelog posts. Can either be a path to a directory or a single Markdown file.

```
USAGE
  $ rdme changelogs PATH --key <value> [--github] [--dryRun]

ARGUMENTS
  PATH  Path to a local Markdown file or folder of Markdown files.

FLAGS
  --dryRun       Runs the command without creating/updating any changelogs in ReadMe. Useful for debugging.
  --github       Create a new GitHub Actions workflow for this command.
  --key=<value>  (required) Project API key

DESCRIPTION
  Sync Markdown files to your ReadMe project as Changelog posts. Can either be a path to a directory or a single
  Markdown file.
```

## `rdme custompages PATH`

Sync Markdown/HTML files to your ReadMe project as Custom Pages. Can either be a path to a directory or a single Markdown/HTML file.

```
USAGE
  $ rdme custompages PATH --key <value> [--github] [--dryRun]

ARGUMENTS
  PATH  Path to a local Markdown file or folder of Markdown files.

FLAGS
  --dryRun       Runs the command without creating/updating any custom pages in ReadMe. Useful for debugging.
  --github       Create a new GitHub Actions workflow for this command.
  --key=<value>  (required) Project API key

DESCRIPTION
  Sync Markdown/HTML files to your ReadMe project as Custom Pages. Can either be a path to a directory or a single
  Markdown/HTML file.
```

## `rdme docs PATH`

Sync Markdown files to your ReadMe project as Guides. Can either be a path to a directory or a single Markdown file.

```
USAGE
  $ rdme docs PATH --key <value> [--version <value>] [--github] [--dryRun]

ARGUMENTS
  PATH  Path to a local Markdown file or folder of Markdown files.

FLAGS
  --dryRun           Runs the command without creating/updating any docs in ReadMe. Useful for debugging.
  --github           Create a new GitHub Actions workflow for this command.
  --key=<value>      (required) Project API key
  --version=<value>  Project version. If running command in a CI environment and this option is not passed, the main
                     project version will be used.

DESCRIPTION
  Sync Markdown files to your ReadMe project as Guides. Can either be a path to a directory or a single Markdown file.

ALIASES
  $ rdme guides
```

## `rdme docs:prune FOLDER`

Delete any docs from ReadMe if their slugs are not found in the target folder.

```
USAGE
  $ rdme docs:prune FOLDER --key <value> [--version <value>] [--github] [--confirm] [--dryRun]

ARGUMENTS
  FOLDER  A local folder containing the files you wish to prune.

FLAGS
  --confirm          Bypass the confirmation prompt. Useful for CI environments.
  --dryRun           Runs the command without deleting any docs in ReadMe. Useful for debugging.
  --github           Create a new GitHub Actions workflow for this command.
  --key=<value>      (required) Project API key
  --version=<value>  Project version. If running command in a CI environment and this option is not passed, the main
                     project version will be used.

DESCRIPTION
  Delete any docs from ReadMe if their slugs are not found in the target folder.

ALIASES
  $ rdme guides:prune
```

## `rdme help [COMMAND]`

Display help for rdme.

```
USAGE
  $ rdme help [COMMAND...] [-n]

ARGUMENTS
  COMMAND...  Command to show help for.

FLAGS
  -n, --nested-commands  Include all nested commands in the output.

DESCRIPTION
  Display help for rdme.
```

_See code: [@oclif/plugin-help](https://github.com/oclif/plugin-help/blob/v6.2.17/src/commands/help.ts)_

## `rdme login`

Login to a ReadMe project.

```
USAGE
  $ rdme login [--email <value>] [--password <value>] [--otp <value>] [--project <value>]

FLAGS
  --email=<value>     Your email address
  --otp=<value>       Your one-time password (if you have two-factor authentication enabled)
  --password=<value>  Your password
  --project=<value>   The subdomain of the project you wish to log into

DESCRIPTION
  Login to a ReadMe project.
```

## `rdme logout`

Logs the currently authenticated user out of ReadMe.

```
USAGE
  $ rdme logout

DESCRIPTION
  Logs the currently authenticated user out of ReadMe.
```

## `rdme open`

Open your current ReadMe project in the browser.

```
USAGE
  $ rdme open [--dash]

FLAGS
  --dash  Opens your current ReadMe project dashboard.

DESCRIPTION
  Open your current ReadMe project in the browser.
```

## `rdme openapi [SPEC]`

Upload, or resync, your OpenAPI/Swagger definition to ReadMe.

```
USAGE
  $ rdme openapi [SPEC] --key <value> [--version <value>] [--id <value>] [--title <value>]
    [--workingDirectory <value>] [--github] [--dryRun] [--useSpecVersion] [--raw] [--create | --update]

ARGUMENTS
  SPEC  A file/URL to your API definition

FLAGS
  --create                    Bypasses the create/update prompt and creates a new API definition in ReadMe.
  --dryRun                    Runs the command without creating/updating any API Definitions in ReadMe. Useful for
                              debugging.
  --github                    Create a new GitHub Actions workflow for this command.
  --id=<value>                Unique identifier for your API definition. Use this if you're re-uploading an existing API
                              definition.
  --key=<value>               (required) Project API key
  --raw                       Return the command results as a JSON object instead of a pretty output.
  --title=<value>             An override value for the `info.title` field in the API definition
  --update                    Bypasses the create/update prompt and automatically updates an existing API definition in
                              ReadMe.
  --useSpecVersion            Uses the version listed in the `info.version` field in the API definition for the project
                              version parameter.
  --version=<value>           Project version. If running command in a CI environment and this option is not passed, the
                              main project version will be used.
  --workingDirectory=<value>  Working directory (for usage with relative external references)

DESCRIPTION
  Upload, or resync, your OpenAPI/Swagger definition to ReadMe.

FLAG DESCRIPTIONS
  --update  Bypasses the create/update prompt and automatically updates an existing API definition in ReadMe.

    Bypasses the create/update prompt and automatically updates an existing API definition in ReadMe. Note that this
    flag only works if there's only one API definition associated with the current version.
```

## `rdme openapi:convert [SPEC]`

Convert an API definition to OpenAPI and bundle any external references.

```
USAGE
  $ rdme openapi:convert [SPEC] [--out <value>] [--workingDirectory <value>]

ARGUMENTS
  SPEC  A file/URL to your API definition

FLAGS
  --out=<value>               Output file path to write converted file to
  --workingDirectory=<value>  Working directory (for usage with relative external references)

DESCRIPTION
  Convert an API definition to OpenAPI and bundle any external references.
```

## `rdme openapi:inspect [SPEC]`

Analyze an OpenAPI/Swagger definition for various OpenAPI and ReadMe feature usage.

```
USAGE
  $ rdme openapi:inspect [SPEC] [--feature
    additionalProperties|callbacks|circularRefs|discriminators|links|style|polymorphism|serverVariables|webhooks|xml|rea
    dme...] [--title <value>] [--workingDirectory <value>]

ARGUMENTS
  SPEC  A file/URL to your API definition

FLAGS
  --feature=<option>...       A specific OpenAPI or ReadMe feature you wish to see detailed information on (if it
                              exists). If any features supplied do not exist within the API definition an exit(1) code
                              will be returned alongside the report.
                              <options: additionalProperties|callbacks|circularRefs|discriminators|links|style|polymorph
                              ism|serverVariables|webhooks|xml|readme>
  --title=<value>             An override value for the `info.title` field in the API definition
  --workingDirectory=<value>  Working directory (for usage with relative external references)

DESCRIPTION
  Analyze an OpenAPI/Swagger definition for various OpenAPI and ReadMe feature usage.
```

## `rdme openapi:reduce [SPEC]`

Reduce an OpenAPI definition into a smaller subset.

```
USAGE
  $ rdme openapi:reduce [SPEC] [--method <value>...] [--out <value>] [--path <value>...] [--tag <value>...] [--title
    <value>] [--workingDirectory <value>]

ARGUMENTS
  SPEC  A file/URL to your API definition

FLAGS
  --method=<value>...         Methods to reduce by (can only be used alongside the `path` option)
  --out=<value>               Output file path to write reduced file to
  --path=<value>...           Paths to reduce by
  --tag=<value>...            Tags to reduce by
  --title=<value>             An override value for the `info.title` field in the API definition
  --workingDirectory=<value>  Working directory (for usage with relative external references)

DESCRIPTION
  Reduce an OpenAPI definition into a smaller subset.
```

## `rdme openapi:validate [SPEC]`

Validate your OpenAPI/Swagger definition.

```
USAGE
  $ rdme openapi:validate [SPEC] [--github] [--workingDirectory <value>]

ARGUMENTS
  SPEC  A file/URL to your API definition

FLAGS
  --github                    Create a new GitHub Actions workflow for this command.
  --workingDirectory=<value>  Working directory (for usage with relative external references)

DESCRIPTION
  Validate your OpenAPI/Swagger definition.

ALIASES
  $ rdme validate
```

## `rdme versions`

List versions available in your project or get a version by SemVer (https://semver.org/).

```
USAGE
  $ rdme versions --key <value> [--version <value>]

FLAGS
  --key=<value>      (required) Project API key
  --version=<value>  A specific project version to view.

DESCRIPTION
  List versions available in your project or get a version by SemVer (https://semver.org/).
```

## `rdme versions:create VERSION`

Create a new version for your project.

```
USAGE
  $ rdme versions:create VERSION --key <value> [--fork <value>] [--codename <value>] [--main true|false] [--beta
    true|false] [--deprecated true|false] [--hidden true|false]

ARGUMENTS
  VERSION  The version you'd like to create. Must be valid SemVer (https://semver.org/)

FLAGS
  --beta=<option>        Should this version be in beta?
                         <options: true|false>
  --codename=<value>     The codename, or nickname, for a particular version.
  --deprecated=<option>  Should this version be deprecated? The main version cannot be deprecated.
                         <options: true|false>
  --fork=<value>         The semantic version which you'd like to fork from.
  --hidden=<option>      Should this version be hidden? The main version cannot be hidden.
                         <options: true|false>
  --key=<value>          (required) Project API key
  --main=<option>        Should this be the main version for your project?
                         <options: true|false>

DESCRIPTION
  Create a new version for your project.
```

## `rdme versions:delete [VERSION]`

Delete a version associated with your ReadMe project.

```
USAGE
  $ rdme versions:delete [VERSION] --key <value>

ARGUMENTS
  VERSION  The version you'd like to delete.

FLAGS
  --key=<value>  (required) Project API key

DESCRIPTION
  Delete a version associated with your ReadMe project.
```

## `rdme versions:update [VERSION]`

Update an existing version for your project.

```
USAGE
  $ rdme versions:update [VERSION] --key <value> [--newVersion <value>] [--codename <value>] [--main true|false]
    [--beta true|false] [--deprecated true|false] [--hidden true|false]

ARGUMENTS
  VERSION  The existing version you'd like to update.

FLAGS
  --beta=<option>        Should this version be in beta?
                         <options: true|false>
  --codename=<value>     The codename, or nickname, for a particular version.
  --deprecated=<option>  Should this version be deprecated? The main version cannot be deprecated.
                         <options: true|false>
  --hidden=<option>      Should this version be hidden? The main version cannot be hidden.
                         <options: true|false>
  --key=<value>          (required) Project API key
  --main=<option>        Should this be the main version for your project?
                         <options: true|false>
  --newVersion=<value>   What should the version be renamed to?

DESCRIPTION
  Update an existing version for your project.
```

## `rdme whoami`

Displays the current user and project authenticated with ReadMe.

```
USAGE
  $ rdme whoami

DESCRIPTION
  Displays the current user and project authenticated with ReadMe.
```
<!-- commandsstop -->
