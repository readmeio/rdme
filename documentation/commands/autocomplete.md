`rdme autocomplete`
===================

Display autocomplete installation instructions.

* [`rdme autocomplete [SHELL]`](#rdme-autocomplete-shell)

## `rdme autocomplete [SHELL]`

Display autocomplete installation instructions.

```
USAGE
  $ rdme autocomplete [SHELL] [-r]

ARGUMENTS
  SHELL  (zsh|bash|powershell) Shell type

FLAGS
  -r, --refresh-cache  Refresh cache (ignores displaying instructions)

DESCRIPTION
  Display autocomplete installation instructions.

EXAMPLES
  $ rdme autocomplete

  $ rdme autocomplete bash

  $ rdme autocomplete zsh

  $ rdme autocomplete powershell

  $ rdme autocomplete --refresh-cache
```

_See code: [@oclif/plugin-autocomplete](https://github.com/oclif/plugin-autocomplete/blob/v3.2.8/src/commands/autocomplete/index.ts)_
