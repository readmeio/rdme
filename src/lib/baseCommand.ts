/* eslint-disable class-methods-use-this */
import type commands from '../cmds';
import type { CommandLineOptions } from 'command-line-args';
import type { OptionDefinition } from 'command-line-usage';

import chalk from 'chalk';

import configstore from './configstore';
import getCurrentConfig from './getCurrentConfig';
import isCI from './isCI';
import { debug, info, warn } from './logger';
import loginFlow from './loginFlow';

export type CommandOptions<T> = T & {
  github?: boolean;
  key?: string;
  version?: string;
} & CommandLineOptions;

export enum CommandCategories {
  ADMIN = 'admin',
  APIS = 'apis',
  CATEGORIES = 'categories',
  CHANGELOGS = 'changelogs',
  CUSTOM_PAGES = 'custompages',
  DOCS = 'docs',
  UTILITIES = 'utilities',
  VERSIONS = 'versions',
}

export default class Command {
  /**
   * The command name
   *
   * @example openapi
   */
  command: keyof typeof commands;

  /**
   * Example command usage, used on invidivual command help screens
   *
   * @example openapi [file] [options]
   */
  usage: string;

  /**
   * The command description, used on help screens
   *
   * @example Upload, or resync, your OpenAPI/Swagger definition to ReadMe.
   */
  description: string;

  /**
   * The category that the command belongs to, used on
   * the general help screen to group commands together
   * and on individual command help screens
   * to show related commands
   *
   * @example CommandCategories.APIS
   */
  cmdCategory: CommandCategories;

  /**
   * Should the command be hidden from our `--help` screens?
   */
  hidden = false;

  /**
   * Does the command run the GitHub Actions onboarding called via
   * `src/index.ts`?
   */
  supportsGHA = false;

  /**
   * Arguments to hide from the individual command help screen
   * (typically used for hiding default arguments)
   *
   * @example ['spec']
   */
  hiddenArgs: string[] = [];

  /**
   * All documented arguments for the command
   */
  args: OptionDefinition[];

  async run(opts: CommandOptions<{}>): Promise<string> {
    Command.debug(`command: ${this.command}`);
    Command.debug(`opts: ${JSON.stringify(opts)}`);

    if (this.args.some(arg => arg.name === 'key')) {
      const { email, project } = getCurrentConfig();

      // We only want to log this if the API key is stored in the configstore, **not** in an env var.
      if (opts.key && configstore.get('apiKey') === opts.key) {
        info(
          `ðŸ”‘ ${chalk.green(email)} is currently logged in, using the stored API key for this project: ${chalk.blue(
            project,
          )}`,
          { includeEmojiPrefix: false },
        );
      }

      if (!opts.key) {
        if (isCI()) {
          throw new Error('No project API key provided. Please use `--key`.');
        }
        info("Looks like you're missing a ReadMe API key, let's fix that! ðŸ¦‰", { includeEmojiPrefix: false });
        const result = await loginFlow();
        info(result, { includeEmojiPrefix: false });
        // eslint-disable-next-line no-param-reassign
        opts.key = configstore.get('apiKey');
      }
    }

    if (opts.github && isCI()) {
      throw new Error('The `--github` flag is only for usage in non-CI environments.');
    }

    // This is a bit of a hack so we can keep our types consistent
    // for this `run` function.
    return Promise.resolve('');
  }

  /**
   * Used in any command where `github` is an option.
   */
  getGitHubArg(): OptionDefinition {
    return {
      name: 'github',
      type: Boolean,
      description: 'Create a new GitHub Actions workflow for this command.',
    };
  }

  /**
   * Used in any command where `key` is an option.
   */
  getKeyArg(): OptionDefinition {
    return {
      name: 'key',
      type: String,
      description: 'Project API key',
    };
  }

  /**
   * Used in the `openapi` family of commands where `title` is an option.
   */
  getTitleArg(): OptionDefinition {
    return {
      name: 'title',
      type: String,
      description: 'An override value for the `info.title` field in the API definition',
    };
  }

  /**
   * Used in any command where `version` is an option.
   */
  getVersionArg(): OptionDefinition {
    return {
      name: 'version',
      type: String,
      description:
        'Project version. If running command in a CI environment and this option is not passed, the main project version will be used.',
    };
  }

  /**
   * Used in the `versions:create` and `versions:update` commands.
   */
  getVersionOpts(): OptionDefinition[] {
    return [
      {
        name: 'version',
        type: String,
        defaultOption: true,
      },
      {
        name: 'codename',
        type: String,
        description: 'The codename, or nickname, for a particular version.',
      },
      {
        name: 'main',
        type: String,
        description:
          "Should this version be the primary (default) version for your project? (Must be 'true' or 'false')",
      },
      {
        name: 'beta',
        type: String,
        description: "Is this version in beta? (Must be 'true' or 'false')",
      },
      {
        name: 'deprecated',
        type: String,
        description: "Would you like to deprecate this version? (Must be 'true' or 'false')",
      },
      {
        name: 'isPublic',
        type: String,
        description:
          "Would you like to make this version public? Any primary version must be public. (Must be 'true' or 'false')",
      },
    ];
  }

  /**
   * Used in the `openapi` family of commands where `workingDirectory` is an option.
   */
  getWorkingDirArg(): OptionDefinition {
    return {
      name: 'workingDirectory',
      type: String,
      description: 'Working directory (for usage with relative external references)',
    };
  }

  static debug(msg: string) {
    debug(msg);
  }

  static info(msg: string) {
    info(msg);
  }

  static warn(msg: string) {
    warn(msg);
  }
}
