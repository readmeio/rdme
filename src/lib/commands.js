const fs = require('fs');
const path = require('path');

exports.load = cmd => {
  let command = cmd;
  let subcommand = '';
  if (cmd.includes(':')) {
    [command, subcommand] = cmd.split(':');
  }

  const file = path.join(__dirname, '../cmds', command, subcommand);
  try {
    // eslint-disable-next-line global-require, import/no-dynamic-require
    const Command = require(file);
    return new Command();
  } catch (e) {
    throw new Error('Command not found.');
  }
};

exports.listByCategory = () => {
  const categories = exports.getCategories();
  const cmds = exports.list();
  cmds.forEach(c => {
    categories[c.command.cmdCategory].commands.push({
      name: c.command.command,
      description: c.command.description,
      position: c.command.position,
    });
  });

  return categories;
};

exports.getSimilar = (cmdCategory, excludeCommand) => {
  const categories = exports.listByCategory();
  return categories[cmdCategory].commands.filter(cmd => cmd.name !== excludeCommand);
};

exports.list = () => {
  const commands = [];
  const cmdDir = `${__dirname}/../cmds`;
  const files = fs
    .readdirSync(cmdDir)
    .map(file => {
      const stats = fs.statSync(path.join(cmdDir, file));
      if (stats.isDirectory()) {
        return fs.readdirSync(path.join(cmdDir, file)).map(f => path.join(file, f));
      }
      return [file];
    })
    .reduce((a, b) => a.concat(b), [])
    .filter(file => file.endsWith('.js'))
    .map(file => path.join(cmdDir, file));

  files.forEach(file => {
    // eslint-disable-next-line global-require, import/no-dynamic-require
    const Command = require(file);

    commands.push({
      file,
      command: new Command(),
    });
  });

  return commands;
};

exports.getCategories = () => {
  return {
    apis: {
      description: 'Upload OpenAPI/Swagger definitions',
      commands: [],
    },
    docs: {
      description: 'Documentation',
      commands: [],
    },
    changelogs: {
      description: 'Changelog',
      commands: [],
    },
    custompages: {
      description: 'Custom Pages',
      commands: [],
    },
    categories: {
      description: 'Categories',
      commands: [],
    },
    versions: {
      description: 'Versions',
      commands: [],
    },
    admin: {
      description: 'Administration',
      commands: [],
    },
    utilities: {
      description: 'Other useful commands',
      commands: [],
    },
  };
};
