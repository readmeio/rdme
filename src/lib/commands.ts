import type { CommandCategories } from './baseCommand';

import commands from '../cmds';

function getCategories(): Record<
  string,
  {
    commands: {
      description: string;
      hidden: boolean;
      name: string;
    }[];
    description: string;
  }
> {
  return {
    apis: {
      description: 'OpenAPI / Swagger',
      commands: [],
    },
    docs: {
      description: 'Docs (a.k.a. Guides)',
      commands: [],
    },
    changelogs: {
      description: 'Changelog',
      commands: [],
    },
    custompages: {
      description: 'Custom Pages',
      commands: [],
    },
    versions: {
      description: 'Versions',
      commands: [],
    },
    categories: {
      description: 'Categories',
      commands: [],
    },
    admin: {
      description: 'Administration',
      commands: [],
    },
    utilities: {
      description: 'Other useful commands',
      commands: [],
    },
  };
}

export function list() {
  return Object.entries(commands).map(([name, Cmd]) => {
    return {
      name,
      command: new Cmd(),
    };
  });
}

export function load(cmd: keyof typeof commands) {
  if (!(cmd in commands)) {
    throw new Error('Command not found.');
  }

  return new commands[cmd]();
}

export function listByCategory() {
  const categories = getCategories();
  const cmds = list();
  cmds.forEach(c => {
    categories[c.command.cmdCategory].commands.push({
      name: c.command.command,
      description: c.command.description,
      hidden: c.command.hidden,
    });
  });

  return categories;
}

export function getSimilar(cmdCategory: CommandCategories, excludeCommand: string) {
  const categories = listByCategory();
  return categories[cmdCategory].commands.filter(cmd => cmd.name !== excludeCommand && !cmd.hidden);
}
