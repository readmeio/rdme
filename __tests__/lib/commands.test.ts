/// <reference types="jest-extended" />
/* eslint-disable jest/no-conditional-expect, jest/no-if */
import type Command from '../../src/lib/baseCommand';

import SingleDocCommand from '../../src/cmds/docs/single';
import { CommandCategories } from '../../src/lib/baseCommand';
import * as commands from '../../src/lib/commands';

describe('utils', () => {
  describe('#list', () => {
    it('should have commands returned', () => {
      expect(commands.list()).not.toHaveLength(0);
    });

    describe('commands', () => {
      it('should be configured properly', () => {
        commands.list().forEach(c => {
          const cmd = c.command;

          expect(cmd.command).not.toBeEmpty();
          expect(cmd.usage).not.toBeEmpty();
          expect(cmd.usage).toStartWith(cmd.command);
          expect(cmd.description).not.toBeEmpty();
          expect(cmd.cmdCategory).not.toBeEmpty();
          expect(cmd.position).toBeNumber();
          expect(cmd.args).toBeArray();
          expect(cmd.run).toBeFunction();

          if (cmd.args.length > 0) {
            cmd.args.forEach(arg => {
              expect(arg.name).not.toBeEmpty();
              expect(arg.type).not.toBeEmpty();
            });
          }
        });
      });

      describe('cli standards', () => {
        describe.each<[string, Command]>(commands.list().map(cmd => [cmd.command.command, cmd.command]))(
          '%s',
          (_, command) => {
            it('should have a description that ends with punctuation', () => {
              const description = command.description.replace('[inactive]', '').replace('[deprecated]', '').trim();
              expect(description).toEndWith('.');
            });

            it('should have standardized argument constructs', () => {
              command.args.forEach(arg => {
                if (arg.name === 'key') {
                  expect(arg.description).toBe('Project API key');
                } else if (arg.name === 'version') {
                  // If `version` is a hidden argument on the command, it won't have a description
                  // so we don't need to bother with this test case.
                  if (Array.isArray(command.hiddenArgs) && command.hiddenArgs.indexOf('version') !== -1) {
                    return;
                  }

                  expect(arg.description).toBe(
                    command.command !== 'versions'
                      ? 'Project version. If running command in a CI environment and this option is not passed, the main project version will be used.'
                      : 'A specific project version to view.'
                  );
                }
              });
            });
          }
        );
      });
    });
  });

  describe('#load', () => {
    it('should load a valid command', () => {
      expect(commands.load('docs:single')).toBeInstanceOf(SingleDocCommand);
    });

    it('should throw an error on an invalid command', () => {
      expect(() => {
        // @ts-expect-error Testing a valid failure case.
        commands.load('buster');
      }).toThrow('Command not found');
    });
  });

  describe('#listByCategory', () => {
    it('should list commands by category', () => {
      expect(commands.listByCategory()).toMatchSnapshot();
    });
  });

  describe('#getSimilar', () => {
    it('should pull similar commands', () => {
      expect(commands.getSimilar(CommandCategories.ADMIN, 'login')).toStrictEqual([
        {
          name: 'logout',
          description: 'Logs the currently authenticated user out of ReadMe.',
          position: 2,
        },
        {
          name: 'whoami',
          description: 'Displays the current user and project authenticated with ReadMe.',
          position: 3,
        },
      ]);
    });
  });
});
