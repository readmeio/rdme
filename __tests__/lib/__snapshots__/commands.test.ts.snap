// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils #listByCategory should list commands by category 1`] = `
Object {
  "admin": Object {
    "commands": Array [
      Object {
        "description": "Login to a ReadMe project.",
        "hidden": false,
        "name": "login",
        "position": 1,
      },
      Object {
        "description": "Logs the currently authenticated user out of ReadMe.",
        "hidden": false,
        "name": "logout",
        "position": 2,
      },
      Object {
        "description": "Displays the current user and project authenticated with ReadMe.",
        "hidden": false,
        "name": "whoami",
        "position": 3,
      },
    ],
    "description": "Administration",
  },
  "apis": Object {
    "commands": Array [
      Object {
        "description": "Upload, or resync, your OpenAPI/Swagger definition to ReadMe.",
        "hidden": false,
        "name": "openapi",
        "position": 1,
      },
      Object {
        "description": "Reduce an OpenAPI definition into a smaller subset.",
        "hidden": false,
        "name": "openapi:reduce",
        "position": 2,
      },
      Object {
        "description": "Validate your OpenAPI/Swagger definition.",
        "hidden": false,
        "name": "openapi:validate",
        "position": 3,
      },
      Object {
        "description": "Alias for \`rdme openapi\`. [deprecated]",
        "hidden": true,
        "name": "swagger",
        "position": 4,
      },
      Object {
        "description": "Alias for \`rdme openapi:validate\` [deprecated].",
        "hidden": true,
        "name": "validate",
        "position": 5,
      },
    ],
    "description": "Upload OpenAPI/Swagger definitions",
  },
  "categories": Object {
    "commands": Array [
      Object {
        "description": "Get all categories in your ReadMe project.",
        "hidden": false,
        "name": "categories",
        "position": 1,
      },
      Object {
        "description": "Create a category with the specified title and guide in your ReadMe project.",
        "hidden": false,
        "name": "categories:create",
        "position": 2,
      },
    ],
    "description": "Categories",
  },
  "changelogs": Object {
    "commands": Array [
      Object {
        "description": "Sync a folder of Markdown files to your ReadMe project as Changelog posts.",
        "hidden": false,
        "name": "changelogs",
        "position": 1,
      },
      Object {
        "description": "Sync a single Markdown file to your ReadMe project as a Changelog post.",
        "hidden": false,
        "name": "changelogs:single",
        "position": 2,
      },
    ],
    "description": "Changelog",
  },
  "custompages": Object {
    "commands": Array [
      Object {
        "description": "Sync a folder of Markdown files to your ReadMe project as Custom Pages.",
        "hidden": false,
        "name": "custompages",
        "position": 1,
      },
      Object {
        "description": "Sync a single Markdown file to your ReadMe project as a Custom Page.",
        "hidden": false,
        "name": "custompages:single",
        "position": 2,
      },
    ],
    "description": "Custom Pages",
  },
  "docs": Object {
    "commands": Array [
      Object {
        "description": "Sync a folder of Markdown files to your ReadMe project.",
        "hidden": false,
        "name": "docs",
        "position": 1,
      },
      Object {
        "description": "Edit a single file from your ReadMe project without saving locally.",
        "hidden": false,
        "name": "docs:edit",
        "position": 2,
      },
      Object {
        "description": "Sync a single Markdown file to your ReadMe project.",
        "hidden": false,
        "name": "docs:single",
        "position": 3,
      },
    ],
    "description": "Documentation",
  },
  "utilities": Object {
    "commands": Array [
      Object {
        "description": "Helpful OpenAPI generation tooling. [inactive]",
        "hidden": true,
        "name": "oas",
        "position": 2,
      },
      Object {
        "description": "Open your current ReadMe project in the browser.",
        "hidden": false,
        "name": "open",
        "position": 1,
      },
    ],
    "description": "Other useful commands",
  },
  "versions": Object {
    "commands": Array [
      Object {
        "description": "List versions available in your project or get a version by SemVer (https://semver.org/).",
        "hidden": false,
        "name": "versions",
        "position": 1,
      },
      Object {
        "description": "Create a new version for your project.",
        "hidden": false,
        "name": "versions:create",
        "position": 2,
      },
      Object {
        "description": "Delete a version associated with your ReadMe project.",
        "hidden": false,
        "name": "versions:delete",
        "position": 4,
      },
      Object {
        "description": "Update an existing version for your project.",
        "hidden": false,
        "name": "versions:update",
        "position": 3,
      },
    ],
    "description": "Versions",
  },
}
`;
